text/0032-react-primary-framework.md

```diff
@@ -0,0 +1,105 @@
+# Making React the Primary Framework
+
+- Start Date: 2021/11/26
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/32
+
+## Decision, including impact on distributions
+
+The app shell should already be developed using React, making React a first-class citizen. As such unnecessary framework-independent packages such as `single-spa` can be removed. Instead, `react-router` will be the central router.
+
+Existing distributions / frontend modules should not be impacted - a compatibility layer will be introduced.
+
+## Reason for decision
+
+Right now every frontend module registers a couple of components (mostly "extensions" and "pages"). Each component will be rendered using a completely new rendering tree. This comes with some significant overhead as compared to reusing a single rendering tree.
+
+Since almost all of our components are React components and since in React its very easy to have a single rendering tree dispatching to multiple DOM nodes (using the `createPortal` API of React) we should move to using only a single React rendering tree. The other advantage is that the framework agnostic approach has problems with otherwise simple things such as prop updates. While this is fully integrated and highly optimized in React, the framework agnostic approach right now makes the update inefficient.
```

Comment:
It would be good to have an empirically grounded estimate of the performance impact of this change—with respect to each of these two inefficiencies.

Reply:
> the framework agnostic approach has problems with otherwise simple things such as prop updates

Are there examples other than prop updates, or can we simplify this?